pkg/csi/service/vanilla/controller.go

```diff
@@ -2274,6 +2267,291 @@ func (c *controller) ListSnapshots(ctx context.Context, req *csi.ListSnapshotsRe
+		if err != nil {
+			// Failed to retrieve information of snapshot.
+			log.Errorf("failed to retrieve snapshot information for volume-id: %s snapshot-id: %s from vCenter %s err: %+v",
+				volID, snapID, vCenterHost, err)
+			return nil, "", err
+		}
+		return snapshots, "", nil
+	} else if volumeID != "" {
+		// Retrieve all snapshots for a volume-id
+		// Fetch vCenterHost & volumeManager for given volume, based on VC configuration
+		vCenterManager = getVCenterManagerForVCenter(ctx, c)
+		vCenterHost, volManager, err = getVCenterAndVolumeManagerForVolumeID(ctx, c, volumeID, volumeInfoService)
+		if err != nil {
+			return nil, csifault.CSIInternalFault, logger.LogNewErrorCodef(log, codes.Internal,
+				"failed to get vCenter/volume manager for volume Id: %q. Error: %v", volumeID, err)
+		}
+		// Check for snapshot support
+		isCnsSnapshotSupported, err := vCenterManager.IsCnsSnapshotSupported(ctx, vCenterHost)
+		if err != nil {
+			return nil, "", logger.LogNewErrorCodef(log, codes.Internal,
+				"failed to check if cns snapshot is supported on VC %s due to error: %v", vCenterHost, err)
+		}
+		if !isCnsSnapshotSupported {
+			return nil, "", logger.LogNewErrorCodef(log, codes.Unimplemented,
+				"VC %s version does not support snapshot operations", vCenterHost)
+		}
+		// Validate that the volume-id is of block volume type.
+		queryFilter := cnstypes.CnsQueryFilter{
+			VolumeIds: []cnstypes.CnsVolumeId{{Id: volumeID}},
+		}
+		querySelection := cnstypes.CnsQuerySelection{
+			Names: []string{string(cnstypes.QuerySelectionNameTypeVolumeType)},
+		}
+		queryResult, err := utils.QueryVolumeUtil(ctx, volManager, queryFilter, &querySelection, true)
+		if err != nil {
+			return nil, "", logger.LogNewErrorCodef(log, codes.Internal,
+				"queryVolumeUtil failed for vCenter %s with err=%+v", vCenterHost, err)
+		}
+
+		if len(queryResult.Volumes) == 0 {
+			return nil, "", logger.LogNewErrorCodef(log, codes.Internal,
+				"volumeID %q not found in QueryVolumeUtil in vCenter %s", volumeID, vCenterHost)
+		}
+		if queryResult.Volumes[0].VolumeType == common.FileVolumeType {
+			return nil, "", logger.LogNewErrorCodef(log, codes.Unimplemented,
+				"ListSnapshot for file volume: %q not supported", volumeID)
+		}
+		// Note: Re-using original Query Snapshot function since we expect snapshots reside in same VC as source vol,
+		//		where pagination gets handled on the query results from single VC by underlying query function.
+		return common.QueryVolumeSnapshotsByVolumeID(ctx, volManager, volumeID, maxEntries)
+	} else {
+		// Fetch all snapshot entries from all vCenters configured and filter them as per request
+		return queryAllVolumeSnapshotsVanilla(ctx, c, token, maxEntries)
+	}
+}
+
+// queryAllVolumeSnapshotsVanilla(): This func fetches all volume snapshots from all vCenters configured
+// and filters the result as per token provided
+func queryAllVolumeSnapshotsVanilla(ctx context.Context, c *controller, token string,
```

Comment:
This is designed for multi vc support, the method name can reflect that.

Reply:
done